ingest/src/main/java/com/microsoft/azure/kusto/ingest/QueuedIngestClient.java

```diff
@@ -146,21 +146,22 @@ public IngestionResult ingestFromFile(FileSourceInfo fileSourceInfo, IngestionPr
             String filePath = fileSourceInfo.getFilePath();
             Ensure.fileExists(filePath);
             CompressionType sourceCompressionType = AzureStorageClient.getCompression(filePath);
-            boolean shouldCompress = AzureStorageClient.shouldCompress(sourceCompressionType, ingestionProperties.getDataFormat());
+            IngestionProperties.DataFormat dataFormat = ingestionProperties.getDataFormat();
+            boolean shouldCompress = IngestClientBase.shouldCompress(sourceCompressionType, dataFormat);
 
             File file = new File(filePath);
             String blobName = genBlobName(
                     file.getName(),
                     ingestionProperties.getDatabaseName(),
                     ingestionProperties.getTableName(),
-                    ingestionProperties.getDataFormat(),
+                    dataFormat.name(), // Used to use an empty string if the DataFormat was empty. Now it can't be empty, with a default of CSV.
```

Comment:
If we want to keep the old behaviour , there's nothing stopping us from returning null in getDataFormat() and sending null here. But I think it's better to default to csv instead of having an "empty" case (The comment reads to me as "this is a bad change, but we have to do it" , but I think it's a good change).

As a bonus, you can annotate `getDataFormat` with `@NotNull` and then remove the null checks in `shouldCompress` and `genBlobName`

Reply:
'The comment reads to me as "this is a bad change, but we have to do it" , but I think it's a good change'
That wasn't my intention - I was just pointing out context for troubleshooting in case it causes unexpected bugs at some point in the future.

I see `shouldCompress()` as more resilient if it can handle a null `DataFormat`, even if it's not currently called that way in our code.

`@NotNull` surprisingly doesn't actually enforce that a parameter isn't null.

Reply:
I guess it's hard to argue against more defensive code.
I think you should add `@NotNull` anyway, just to make sure we don't accidentally change the contract

Reply:
I just don't understand what that would achieve.

The javadoc for the method already specifies that it doesn't return null:
![image](https://user-images.githubusercontent.com/30245027/147051573-30f37368-7371-48e0-9d32-a3bf157a3117.png)

Reply:
You'd get better warnings, it solidifies the contract.

Reply:
I think they might only see a warning if they're using an IDE with (certain types of?) lint checking.

But why do we even want to add a (not useful, IMHO) `@NotNull` annotation to a method that explicitly handles `NULL`?

Reply:
@NotNull is on the return type, we guarantee that the method doesn't return null

Reply:
Then I don't think that would impact any warnings in code using this method - I don't recall adding `@NotNull` to the return type doing that in my experience. I think it could only produce a warning if someone had this method return null (assuming they're using an IDE with (certain types of?) lint checking), but that's already clear in the javadoc that it doesn't. I think the annotation on its own (without explicit code to enforce) is only to document.

I think the question here boils down to how we want to manage API contract documentation -
1) Javadoc
2) Annotations, like `@NotNull`
3) Both
4) Something else

I'm flexible on this, but my thoughts are:
- We should be using javadoc no matter what, as that's the standard for documenting APIs.
- I prefer one source of truth, meaning I prefer not using annotations for documentation.
- I see these annotations as less beneficial for return types as for input parameters, since it helps the **user** of the method.
- I do indeed see a benefit to these annotations, and have incorporated them into several (most?) projects I've built. But that's been when I both (a) use a framework that enforces them as well (like IIRC Spring provides) and not just a coder-side IDE lint check, and (b) do it consistently throughout the codebase.

Actually, I'm looking at the code now to see whether we use it, and in fact we do about 40 times - and it looks like I wrote most of them! Mostly on input parameters, but a few on the return type. So I'm a bit embarrassed :-) but it shows that I'm very much into it, though inconsistently - meaning it's good to have this discussion of how we want to manage API contract documentation.

Reply:
I feel the extra security from annotations is worth it, even if it's just a lint check.
I feel like the annotations are part of the function signature, and they are the "stronger" source of truths, like types.

But to be fair, @NotNull is much less valuable the @CanBeNull, since the ladder actually stops NREs.
So I trust you here, you can decide if you want to annotate or not.